app/service/data_svc.py

```diff
@@ -19,6 +19,25 @@
 from app.utility.base_service import BaseService
 
 Adjustment = namedtuple('Adjustment', 'ability_id trait value offset')
+history = defaultdict(set)
+
+
+def track_recursion(f):
+    """
+    decorator to track recursive calls in stateless functions to prevent infinite loops
+    """
+    function_history = history[f.__name__]
+
+    async def call(*args, **kwargs):
+        key = str({'args': args, 'kwargs': kwargs})
+        if key in function_history:
+            function_history.clear()
+            raise Exception(f'infinite loop detected, function:{f.__name__}, inputs:{key}')
+        function_history.add(key)
+        ret = await f(*args, **kwargs)
+        function_history.remove(key)
```

Comment:
since we have an async application, there's a good likelihood we'll have multiple "threads" entering here at once. how resilient will this design be?

Reply:
currently we would not have that issue since all async calls decorated by this would be done synchronously as they all funnel from load_adversaries, but if we were to branch out and use it on other functions that could be called from multiple places concurrently, then we may have an issue with the global nature of the `history` defaultdict we are using for the function name/params history, in which case a class based decorator would be a better solution to this, but that then reduces the recursion depth you can go (from ~500 with this implementation to ~333 with a class based, and all that is down from the 1k you get normally)

Reply:
`class FancyCrumbs:
    def __init__(self, function):
        self.f = function
        self.history = set()

    async def __call__(self, *args, **kwargs):
        key = str({'args': args, 'kwargs': kwargs})
        if key in self.history:
            raise Exception(f'we are in a loop, function:{self.f.__name__}, inputs:{key}')
        self.history.add(key)
        ret = await self.f(*args, **kwargs)
        self.history.remove(key)
        return ret`